archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,576 @@
+(sphere.second_inter_mem _).2 cfg.B_mem_circumsphere
+
+lemma symm_A₂ : cfg.symm.A₂ = cfg.B₂ :=
+by { simp_rw [A₂, B₂, symm_triangle_ABC_circumsphere], refl }
+
+lemma QP_parallel_BA : line[ℝ, cfg.Q, cfg.P] ∥ line[ℝ, cfg.B, cfg.A] :=
+by { rw [set.pair_comm cfg.Q, set.pair_comm cfg.B], exact cfg.PQ_parallel_AB }
+
+lemma A_ne_A₁ : cfg.A ≠ cfg.A₁ :=
+begin
+  intro h,
+  have h' := cfg.not_collinear_ABC,
+  rw [h, set.insert_comm] at h',
+  exact h' cfg.wbtw_B_A₁_C.collinear
+end
+
+lemma collinear_PAA₁A₂ : collinear ℝ ({cfg.P, cfg.A, cfg.A₁, cfg.A₂} : set Pt) :=
+begin
+  rw [A₂,
+      (cfg.triangle_ABC.circumsphere.second_inter_collinear cfg.A cfg.A₁).collinear_insert_iff_of_ne
+        (set.mem_insert _ _) (set.mem_insert_of_mem _ (set.mem_insert _ _)) cfg.A_ne_A₁,
+      set.insert_comm],
+  exact cfg.wbtw_A_P_A₁.collinear
+end
+
+lemma A₁_ne_C : cfg.A₁ ≠ cfg.C :=
+begin
+  intro h,
+  have hsbtw := cfg.sbtw_Q_A₁_Q₁,
+  rw h at hsbtw,
+  have ha := hsbtw.angle₂₃₁_eq_zero,
+  rw [angle_CQ₁Q_eq_angle_CBA, angle_eq_zero_iff_ne_and_wbtw] at ha,
+  have ht := cfg.not_collinear_ABC,
+  rcases ha with ⟨-, hw⟩ | ⟨-, hw⟩,
+  { have hc := hw.collinear,
+    rw [set.pair_comm, set.insert_comm] at hc,
+    exact ht hc },
+  { have hc := hw.collinear,
+    rw set.insert_comm at hc,
+    exact ht hc }
+end
+
+lemma B₁_ne_C : cfg.B₁ ≠ cfg.C := cfg.symm.A₁_ne_C
+
+lemma Q_not_mem_CB : cfg.Q ∉ line[ℝ, cfg.C, cfg.B] :=
+begin
+  intro hQ,
+  have hQA₁ : line[ℝ, cfg.Q, cfg.A₁] ≤ line[ℝ, cfg.C, cfg.B] :=
+    affine_span_pair_le_of_mem_of_mem hQ cfg.wbtw_B_A₁_C.symm.mem_affine_span,
+  have hQ₁ : cfg.Q₁ ∈ line[ℝ, cfg.C, cfg.B],
+  { rw affine_subspace.le_def' at hQA₁,
+    exact hQA₁ _ cfg.sbtw_Q_A₁_Q₁.right_mem_affine_span },
+  have hc : collinear ℝ ({cfg.C, cfg.Q₁, cfg.Q} : set Pt),
+  { have hc' : collinear ℝ ({cfg.B, cfg.C, cfg.Q₁, cfg.Q} : set Pt),
+    { rw [set.insert_comm cfg.B, set.insert_comm cfg.B, set.pair_comm, set.insert_comm cfg.C,
+          set.insert_comm cfg.C],
+      exact collinear_insert_insert_of_mem_affine_span_pair hQ₁ hQ },
+    exact hc'.subset (set.subset_insert _ _) },
+  rw [collinear_iff_eq_or_eq_or_angle_eq_zero_or_angle_eq_pi, cfg.angle_CQ₁Q_eq_angle_CBA,
+      or_iff_right cfg.C_ne_Q₁, or_iff_right cfg.sbtw_Q_A₁_Q₁.left_ne_right] at hc,
+  have hca : collinear ℝ ({cfg.A, cfg.B, cfg.C} : set Pt),
+  { rwa [collinear_iff_eq_or_eq_or_angle_eq_zero_or_angle_eq_pi, or_iff_right cfg.B_ne_C.symm,
+         or_iff_right cfg.A_ne_B, angle_comm] },
+  exact cfg.not_collinear_ABC hca
+end
+
+lemma Q_ne_B : cfg.Q ≠ cfg.B :=
+begin
+  intro h,
+  have h' := cfg.Q_not_mem_CB,
+  rw h at h',
+  exact h' (right_mem_affine_span_pair _ _ _)
+end
+
+lemma s_opp_side_CB_Q_Q₁ : line[ℝ, cfg.C, cfg.B].s_opp_side cfg.Q cfg.Q₁ :=
+cfg.sbtw_Q_A₁_Q₁.s_opp_side_of_not_mem_of_mem cfg.Q_not_mem_CB cfg.wbtw_B_A₁_C.symm.mem_affine_span
+
+section oriented
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma oangle_CQ₁Q_sign_eq_oangle_CBA_sign :
```

Comment:
One interpretation of the original problem is that `∡ cfg.C cfg.Q₁ cfg.Q = ∡ cfg.C cfg.B cfg.A`. Is this not how you read the question? Are angles by default not oriented? I guess they're not.

Reply:
I think the correct interpretation of the informal statement is in terms of unoriented angles, and angles should not be considered oriented by default. (And certainly should not be considered oriented mod pi by default, even though that's often convenient in proofs, because you can definitely talk about `sin` and `cos` of informal angles, and you can do that mod pi.)